openmdao/core/group.py

```diff
@@ -195,12 +352,72 @@ def _setup_procs(self, pathname, comm):
         # Compute _subsystems_proc_range
         self._subsystems_proc_range = [sub_proc_range] * len(self._subsystems_myproc)
 
+        self._local_system_set = set()
+
         # Perform recursion
         for subsys in self._subsystems_myproc:
+            subsys._local_vector_class = self._local_vector_class
+            subsys._distributed_vector_class = self._distributed_vector_class
+
             if self.pathname:
-                subsys._setup_procs('.'.join((self.pathname, subsys.name)), sub_comm)
+                subsys._setup_procs('.'.join((self.pathname, subsys.name)), sub_comm, mode)
             else:
-                subsys._setup_procs(subsys.name, sub_comm)
+                subsys._setup_procs(subsys.name, sub_comm, mode)
+
+        # build a list of local subgroups to speed up later loops
+        self._subgroups_myproc = [s for s in self._subsystems_myproc if isinstance(s, Group)]
+
+    def _check_reconf_update(self):
+        """
+        Check if any subsystem has reconfigured and if so, perform the necessary update setup.
+        """
+        # See if any local subsystem has reconfigured
+        reconf = np.any([subsys._reconfigured for subsys in self._subsystems_myproc])
+
+        # See if any subsystem on this or any other processor has configured
+        if self.comm.size > 1:
+            reconf = self.comm.allreduce(reconf) > 0
+
+        if reconf:
+            # Perform an update setup
+            with self._unscaled_context_all():
+                self.resetup('update')
+
+            # Reset the _reconfigured attribute to False
+            for subsys in self._subsystems_myproc:
+                subsys._reconfigured = False
+
+            self._reconfigured = True
+
+    def _list_states(self):
```

Comment:
I don't disagree.  I simply moved these functions from System to Group but know nothing about them really.  I did look at them a little based on your comment, and it turns out that _list_states_allprocs() isn't correct.  It actually won't lists states from all processes.  In fact I'd expect it to raise an exception when it calls _list_states_allprocs on a non-local subsystem.  Not sure what to do about it here though since it has nothing really to do with this story.  Maybe add a story to fix it?

Reply:
Ok I went ahead and fixed _list_states_allprocs and added a test for it.